Use pivot_root in addition to chroot when possible

hvdijk opened the pull request:
As brought up [earlier on the mailing list](http://lists.science.uu.nl/pipermail/nix-dev/2014-December/015089.html), here is a request to make Nix use pivot_root on Linux. I missed that the sandbox support mentioned in that thread that I was waiting for already got accepted in Nix. There does not appear to be any overlap with my proposed change.

chroot only changes the process root directory, not the mount namespace root directory, and it is well-known that any process with chroot capability can break out of a chroot "jail". By using pivot_root as well, and unmounting the original mount namespace root directory, breaking out becomes impossible.

Non-root processes typically have no ability to use chroot() anyway, but they can gain that capability through the use of clone() or unshare(). For security reasons, these syscalls are limited in functionality when used inside a normal chroot environment. Using pivot_root() this way does allow those syscalls to be put to their full use.

I've tested this change with my own packages as well as nixpkgs, inside a chroot and outside. For nixpkgs, there are no issues. It doesn't matter whether Nix itself is running in a chroot, the packages in nixpkgs build without any issues. For my own packages, as long as Nix itself is not run in a chroot, clone()/unshare() are fully functional. If Nix itself is running in a chroot, this is not sufficient to get them to work, but that is not something I'm trying to make work either.

On the mailing list, an alternative suggestion of the equivalent of mount --move was brought up. I gave that a shot, but couldn't get it working. After mount-moving /dir to /, / still referred to the original root in my testing, and following that by chroot (if the current directory was /dir) was insufficient to make clone()/unshare() work the way I wanted them to.

shlevy commented:
"After mount-moving /dir to /, / still referred to the original root in my testing" <- Did you verify that `/` was not mounted shared? Can you provide a repro of this behavior? That seems definitely broken.

edolstra commented:
Is it actually useful to make this configurable? It doesn't sound like pivot_root would break anything, so in that case we should just do it unconditionally.

hvdijk commented:
@shlevy: I will attempt to create a stand-alone test case.
@edolstra: The second code path needs to be kept for systems that do have chroot(), but do not have pivot_root(), so I think it needs to be compile-time configurable at the least. And if there are two separate code paths, then making it run-time configurable allows both of those code paths to be tested on the same system. But I agree that for most users, it doesn't make sense to disable the option, that's why I enabled it by default.

edolstra commented:
Hm, what systems don't have `pivot_root`? The manpage says it was introduced in Linux 2.3.41, a long time ago...

hvdijk commented:
chroot() is available on many systems, but pivot_root() is specific to Linux. Although, looking closer, it seems like the current chroot() facility actually isn't used or useful outside of Linux anyway, so in that case, yeah, I have no objections to making it unconditional.

edolstra commented:
Right, the chroot support in Nix is totally Linux-specific.

hvdijk commented:
@shlevy: https://gist.github.com/hvdijk/e3a6443ac40574331790 is a stand-alone test that shows the behaviour I describe. system("ls /") is unaffected by the mount move from /bin to /. I did the equivalent of mount --make-rprivate / too, to be sure the problem is not what you said. If I'm doing something stupid, please feel free to point it out.

@edolstra I've removed all the unnecessary bits, updated it to always use pivot_root. I've tested that it compiles, not much more, but I'll be using this version myself starting today, so if there are any problems I should spot them quickly. I do see an existing `#ifdef __linux__` check inside the `CHROOT_ENABLED` block, I guess that is unnecessary then?

shlevy commented:
@hvdijk not a computer right now, what happens if you `chdir("/")`` before the `ls`?

hvdijk commented:
@shlevy No change. `chdir("/")` succeeds, but it changes to the old root. It does so even within that one program, as can be seen from the fact that a subsequent `chdir("bin")` also works.

shlevy commented:
Very strange! That behavior seems inconsistent. Guess `pivot_root` is the best way...

hvdijk commented:
I've done some testing now. I switched nixpkgs from master to release-14.12 to force a mass rebuild, which completed without any issues other than the libsigsegv one, which is unrelated. My own packages are happy with the current version as well.

hvdijk commented:
And I see that [it was accepted](https://github.com/NixOS/nix/commits/master). This can be closed now. Thanks!
